### PR TITLE
Island: Change schema compiler to merge plugins and exploiters

### PR DESCRIPTION
**Pull request by VakarisZ**
# What does this PR do?

Exploiter plugins and hard-coded plugins must be merged to create a cohesive configuration schema

Fixes part of #2785

## PR Checklist
* [ ] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [ ] Added relevant unit tests?
* [ ] Have you successfully tested your changes locally? Elaborate:
    > Tested by running unit tests
* [ ] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/monkey_island/cc/repositories/utils/hard_coded_exploiters.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        "title": "Hadoop Exploiter",
+        "info": "Exploits a privilege escalation vulnerability (CVE-2020-1472) in a Windows "
+        + "server domain controller (DC) by using the Netlogon Remote Protocol (MS-NRPC). "
+        + "This exploiter changes the password of a Windows server DC account, steals "
+        + "credentials, and then attempts to restore the original DC password. The victim "
+        "DC "
+        + "will be unable to communicate with other DCs until the original "
+        + "password has been restored. If Infection Monkey fails to restore the "
+        + 'password automatically, you"ll have to do it manually. For more '
+        + "information, see the documentation.",
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/repositories/utils/hard_coded_exploiters.py:13]: Newlines here need to be fixed

File: monkey/monkey_island/cc/repositories/utils/hard_coded_exploiters.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+HARD_CODED_EXPLOITER_PLUGINS = {
+    "HadoopExploiter": {
+        "type": "object",
+        "title": "Hadoop Exploiter",
+        "info": "Exploits a privilege escalation vulnerability (CVE-2020-1472) in a Windows "
+        "server domain controller (DC) by using the Netlogon Remote Protocol (MS-NRPC). "
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/repositories/utils/hard_coded_exploiters.py:6]: This is the wrong description, will fix in later PR

File: monkey/monkey_island/cc/repositories/utils/agent_configuration_schema_compiler.py [hunk 3]
```diff
@@ -61,15 +39,25 @@ def _add_plugins(self, schema: Dict[str, Any]) -> Dict[str, Any]:
...
+    def _add_properties_to_objects(self, schema: Dict[str, Any]) -> Dict[str, Any]:
+        for plugin_path in PLUGIN_PATH_IN_SCHEMA.values():
+            plugin_schema = dpath.util.get(schema, plugin_path, ".")
+            plugin_schema["properties"] = {}
+            plugin_schema["additionalProperties"] = False
+        return schema
+
+    # Exploiters that are not plugins need to be added manually. This should be removed
+    # once all exploiters are turned into plugins
+    def _add_exploiters(self, schema: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repositories/utils/agent_configuration_schema_compiler.py:51]:
```suggestion
    def _add_non_plugin_exploiters(self, schema: Dict[str, Any]) -> Dict[str, Any]:
```

File: monkey/monkey_island/cc/repositories/utils/agent_configuration_schema_compiler.py [hunk 4]
```diff
@@ -61,15 +39,25 @@ def _add_plugins(self, schema: Dict[str, Any]) -> Dict[str, Any]:
 
         return schema
 
-    @staticmethod
-    def _get_plugin_type_string(plugin_type: AgentPluginType):
-        return plugin_type.name.lower()
+    def _add_properties_to_objects(self, schema: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repositories/utils/agent_configuration_schema_compiler.py:42]:
```suggestion
    def _add_properties_field_to_plugin_types(self, schema: Dict[str, Any]) -> Dict[str, Any]:
```

^^ Not sure about this name, but something more descriptive would be good.

**Review by mssalvatore — APPROVED** [review]
Approved, but I think we can improve a few names.

**Pull request merged**